src/control/cmd/daos_agent/infocache.go

```diff
@@ -8,176 +8,445 @@ package main
-func (c *localFabricCache) IsEnabled() bool {
+// IsAttachInfoEnabled checks whether the GetAttachInfo cache is enabled.
+func (c *InfoCache) IsAttachInfoCacheEnabled() bool {
 	if c == nil {
 		return false
 	}
+	return !c.attachInfoCacheDisabled.Load()
+}
+
+// DisableAttachInfoCache fully disables the attach info cache.
+func (c *InfoCache) DisableAttachInfoCache() {
+	if c == nil {
+		return
+	}
+	c.attachInfoCacheDisabled.Store(true)
+}
 
-	return c.enabled.IsTrue()
+// EnableAttachInfoCache enables a refreshable GetAttachInfo cache.
+func (c *InfoCache) EnableAttachInfoCache(interval time.Duration) {
+	if c == nil {
+		return
+	}
+	c.attachInfoRefresh = interval
+	c.attachInfoCacheDisabled.Store(false)
 }
 
-// IsCached reports whether there is data in the cache.
-func (c *localFabricCache) IsCached() bool {
+// IsFabricCacheEnabled checks whether the NUMAFabric cache is enabled.
+func (c *InfoCache) IsFabricCacheEnabled() bool {
 	if c == nil {
 		return false
 	}
+	return !c.fabricCacheDisabled.Load()
+}
+
+// DisableFabricCache fully disables the fabric device cache.
+func (c *InfoCache) DisableFabricCache() {
+	if c == nil {
+		return
+	}
+	c.fabricCacheDisabled.Store(true)
+}
 
-	return c.initialized.IsTrue()
+// EnableFabricCache enables a refreshable local fabric cache.
+func (c *InfoCache) EnableFabricCache() {
+	if c == nil {
+		return
+	}
+	c.fabricCacheDisabled.Store(false)
 }
 
-// Cache caches the results of a fabric scan locally.
-func (c *localFabricCache) CacheScan(ctx context.Context, scan *hardware.FabricInterfaceSet) {
+// EnableStaticFabricCache sets up a fabric cache based on a static value that cannot be refreshed.
+func (c *InfoCache) EnableStaticFabricCache(ctx context.Context, nf *NUMAFabric) {
 	if c == nil {
 		return
 	}
 
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
+	item := &cachedFabricInfo{
+		cacheItem: cacheItem{
+			lastCached: time.Now(),
+		},
+		fetch: func(context.Context, ...string) (*NUMAFabric, error) {
+			return nf, nil
+		},
+		lastResults: nf,
+	}
+	if err := c.cache.Set(item); err != nil {
+		c.log.Errorf("error setting static fabric cache: %v", err)
+	}
+	c.EnableFabricCache()
+}
+
+// GetAttachInfo fetches the attach info from the cache, and refreshes if necessary.
+func (c *InfoCache) GetAttachInfo(ctx context.Context, sys string) (*control.GetAttachInfoResp, error) {
+	if c == nil {
+		return nil, errors.New("InfoCache is nil")
+	}
 
-	scanResult := NUMAFabricFromScan(ctx, c.log, scan)
-	c.setCache(scanResult)
+	if !c.IsAttachInfoCacheEnabled() {
+		return c.getAttachInfoRemote(ctx, sys)
+	}
+
+	// Use the default system if none is specified.
+	if sys == "" {
+		sys = build.DefaultSystemName
+	}
+	createItem := func() (cache.Item, error) {
+		c.log.Debugf("cache miss for %s", sysAttachInfoKey(sys))
+		return newCachedAttachInfo(c.attachInfoRefresh, sys, c.client, c.getAttachInfo), nil
+	}
+
+	item, release, err := c.cache.GetOrCreate(ctx, sysAttachInfoKey(sys), createItem)
+	defer release()
```

Comment:
From what I understand, the item lock is needed to safely copy its content at line 326.
However, as it is not a deep copy, the content of cai.lastResponse could be shared by a different processes after the function return ?
I was expecting to use value instead of pointer to avoid this issue.
However, I am probably missing something as I am not a GO expert.

Reply:
Hm, at first I was thinking that this concern wasn't relevant, since there is a single user of the InfoCache. But as I consider it, there are potentially issues with several clients requesting GetAttachInfo at the same time. Depending on the system hardware, each of those clients might actually be assigned a different network interface. With the pointer passed back, this could be a data race. I'd like to address this in a PR to master, which will eventually trickle back to this branch.

Reply:
OK for me